#### CONTACT INFORMATION
Chin Jian (CJ) Yang  
cjyang90@uwalumni.com  
[cjyang-work.github.io](https://cjyang-work.github.io)  
*Mailing address and phone number are available on request.*


#### SUMMARY STATEMENT

I am a quantitative geneticist with a background in evolutionary and molecular genetics. My current research work focuses on computational analyses in crop data using statistics and simulations. I am looking to apply my skills and experience to take on new challenges in an interdisciplinary role.  

#### EDUCATION

University of Wisconsin-Madison, WI, USA  
PhD Genetics, Aug 2012 - May 2018

Indiana University Bloomington, IN, USA  
BSc Biotechnology and Mathematics, Aug 2009 - May 2012

#### RESEARCH EXPERIENCE

2019 - present: Postdoctoral research associate  
*Advisors*: Dr Wayne Powell, Dr Ian Mackay  
*Topic*: **Development of novel statistical methodologies for modern breeding program improvement**. I partitioned the genomic prediction equation according to  parental contributions for better introgression of exotic alleles into elite breeding populations. I highlighted issues in the current Plant Variety Rights (PVR) system from a genetic perspective and proposed an improved PVR system using molecular markers. I analyzed crossing schemes in multi-parental populations and developed an R package, as well as a Shiny app, for *a priori* testing of population designs via simulations. I showed a trait-free approach for mapping genomic loci under selection in elite crop varieties by modelling allele frequency changes under a logistic distribution. Currently, I am developing a novel approach for mapping variance QTLs (vQTLs) as a proxy for interactions due to epistasis and genetic-by-environment.  

2018 - 2019: Postdoctoral research associate  
*Advisor*: Dr Chris Carolin-Schoen  
*Topic*: **Partitioning genetic correlations of early development traits in maize**. I compared the performances of REML and Bayesian approaches in estimating trait genetic correlations across genomic windows.

2012 - 2018: Graduate research assistant  
*Advisor*: Dr John F Doebley  
*Topic*: **Dissection of the genetic architecture of domestication traits in maize and its ancestor teosinte**. I identified and characterized three important domestication genes that shaped the ear architecture in maize using QTL fine-mapping and molecular biology analyses. I developed a QTL linkage mapping pipeline in the teosinte nested association mapping (TeoNAM) population for undergraduate research training. I modelled the selection strengths and constraints in 18 traits during the initial maize domestication using genetic variances estimated from large proxy populations of ancient maize and teosinte.  

2010 - 2012: Undergraduate research assistant  
*Advisor*: Dr Jeffrey D Palmer  
*Topic*: **Phylogenetic analysis of mitochondrial genes in Lobeliaceae**. I sequenced several mitochondrial genes across hundreds of species in Lobeliaceae (family of small flowering plants) for the use in phylogenetic analyses.  

#### SKILLS

* *Quantitative/Population Genetics*: selection, breeding, genetic gain, genetic variances/covariances, heritabilities, epistasis, phylogenetics.  
* *Statistics*: linear regression, mixed model, REML, Bayesian, genomic selection, GWAS, QTL mapping.  
* *Computational Tools*: LINUX, UNIX, Git, Windows, Mac, Microsoft Office, TASSEL.  
* *Programming*: R, Shiny, Markdown, HTML, Python, SAS, ASReml, Julia.  
* *Big Data*: Genotype-by-Sequencing (GBS), genomic array.  
* *Molecular Genetics*: DNA/RNA isolation, marker development, cloning, PCR, qRT-PCR, gene expression.  
* *Field Work*: field trials, greenhouse trials, phenotyping, travel for work.  
* *Management*: undergraduate mentorship, collaboration.  

#### INVITED TALK

* "MAGIC population design" at the Rank Symposium on Meiosis: Controlling the Recombination Landscape, Grasmere, UK, 4 - 7 April 2022.  

#### ACCEPTED TALK

* "Defining the causes and consequences of phantom epistasis in heterotic hybrids" at the 18th Eucarpia Biometrics in Plant Breeding Conference, Paris, France, 21 - 23 September 2022.  

#### PEER REVIEWS

* Theoretical and Applied Genetics (3)  
* Nature Plants (1)  
* Journal of Experimental Botany (1)  
* Plant Ecology & Diversity (1)  


#### PhD EXAMINATION
* Internal PhD examiner for Jon Bancic (University of Edinburgh) on "Breeding Strategies for Integrating New and Orphan Crops in Production
Systems", 22 November 2021.  

#### FUNDED GRANTS

* "MAGS: Mutation Breeding for Quantitative Traits with Genomic Selection" from the International Atomic Energy Agency (IAEA), 2022 - 2027, Co-Investigator.  
* "BARGAIN: Mobilising untapped genetic diversity, JHI-B1-2" from the Scottish Government's Rural and Environmental Science and Analytical Services (RESAS), 2022 - 2027, Named Staff.   
* "The impact of novel crops and farming practices on the Scottish agricultural landscape, JHI-B1-3" from RESAS, 2022 - 2027, Named Staff.  


#### TEACHING EXPERIENCE

* Guest lecture on "IBD, IBS, Genetic Distance, Population Structure" for the XXIII International Master in Plant Genetics, Genomics and Breeding by CIHEAM Zaragoza, Online, 15 - 19 February 2021.  
* Teaching assistant for General Genetics (code 466) at University of Wisconsin-Madison, Fall 2013.  

#### CONFERENCES AND WORKSHOPS

* Maize Genetics Conference, various locations, 2013, 2015 - 2017, 2020 - 2022.  
* Barley Away Days, Dunkeld, UK, 5 - 6 February 2020.  
* A Century of Genetics, Edinburgh, UK, 13 - 15 November 2019.  
* Forestry Workshop on Genomic Selection, Edinburgh, UK, 17 September 2019.  
* MAGIC Workshop, Cambridge, UK, 23 - 24 July 2019.  
* UW-Madison PSGSC Annual Plant Science Symposium, Madison, WI, 2012 - 2017.  
* 5th International Conference on Quantitative Genetics, Madison, WI, 12 - 17 June 2016.  

#### FELLOWSHIPS AND AWARDS

* 2017: Schlimgen Award; Department of Genetics, University of Wisconsin-Madison.  
* 2016 - 2017: Stone Travel Award; Department of Genetics, University of Wisconsin-Madison.  
* 2007 - 2012: National Scholarship; Public Service Department of Malaysia.  

#### LANGUAGES

* English: native/bilingual proficiency.  
* Malay: limited working proficiency.  
* Mandarin: limited working proficiency.  

#### PUBLICATIONS

1. **Yang CJ**, Ladejobi O, Mott R, Powell W & Mackay I (2022) Analysis of historical selection in winter wheat. Theor Appl Genet. doi.org/10.1007/s00122-022-04163-3.  
2. Samayoa LF, Olukolu BA, **Yang CJ**, Chen Q, Stetter MG, York AM, Sanchez-Gonzalez JJ, Glaubitz JC, Bradbury PJ, Romay MC, Sun Q, Yang J, Ross-Ibarra J, Buckler ES, Doebley JF, Holland JB (2021) Domestication reshaped the genetic basis of inbreeding depression in a maize landrace compared to its wild relative, teosinte. PLoS Genet 17: e1009797.  
3. **Yang CJ**, Edmondson RN, Piepho H-P, Powell W & Mackay I (2021) Crafting for a better MAGIC: systematic design and test for Multiparental Advanced Generation Inter-Cross population. G3 11: jkab295.  
4. Chen Q, Samayoa LF, **Yang CJ**, Olukolu BA, York AM, Sanchez-Gonzalez JJ, Xue W, Glaubitz JC, Bradbury PJ, Romay MC, Sun Q, Buckler ES, Holland JB & Doebely JF (2021) A conserved genetic architecture among populations of the maize progenitor, teosinte, was radically altered by domestication. Proc Natl Acad Sci USA 118: e2112970118.  
5. **Yang CJ**, Russell J, Ramsay L, Thomas W, Powell W & Mackay I (2021) Overcoming barriers to the registration of new plant varieties under the DUS system. Commun Biol 4: 302.  
6. **Yang CJ**, Sharma R, Gorjanc G, Hearne S, Powell W & Mackay I (2020) Origin specific genomic selection: a simple process to optimize the favorable contribution of parents to progeny. G3 10: 2445-2455.  
7. Chen Q, Samayoa LF, **Yang CJ**, Bradbury PJ, Olukolu BA, Neumeyer MA, Romay MC, Sun Q, Lorant A, Buckler ES, Ross-Ibarra J, Holland JB & Doebley JF (2020) The genetic architecture of the maize progenitor, teosinte, and how it was altered during maize domestication. PLoS Genet 16: e1008791.  
8. Chen Q, **Yang CJ**, York AM, Xue W, Daskalska LL, DeValk CA, Krueger KW, Lawton SB, Spiegelberg BG, Schnell JM, Neumeyer MA, Perry JS, Peterson AC, Kim B, Bergstrom L, Yang L, Barber IC, Tian F & Doebley JF (2019) TeoNAM: A nested assocation mapping population for domestication and agronomic trait analysis in maize. Genetics 213: 1065-1078.  
9. **Yang CJ**, Samayoa LF, Bradbury PJ, Olukolu BA, Xue W, York AM, Tuholski MR, Wang W, Daskalska LL, Neumeyer MA, Sanchez-Gonzalez JJ, Romay MC, Glaubitz JC, Sun Q, Buckler ES, Holland JB & Doebley JF (2019) The genetic architecture of teosinte catalyzed and constrained maize domestication. Proc Natl Acad Sci USA 116: 5643-5652.  
10. Guo L, Wang X, Zhao M, Huang C, Li C, Li D, **Yang CJ**, York AM, Xue W, Xu G, Liang Y, Chen Q, Doebley JF & Tian F (2018) Stepwise cis-regulatory changes in *ZCN8* contribute to maize flowering-time adaptation. Curr Biol 28: 3005-2015.  
11. **Yang CJ**, Kursel LE, Studer AJ, Bartlett ME, Whipple CJ & Doebley JF (2016) A gene for genetic background in *Zea mays*: fine-mapping *enhancer of teosinte branched1.2* to a YABBY class transcription factor. Genetics 204: 1573-1585.  
12. Yang L, **Yang CJ**, Cheng Q, Xue W & Doebley JF (2016) Mapping prolificacy QTL in maize and teosinte. J Hered 107: 674-678.  
