---
layout: post
title: Maize Domestication
---

Maize (*Zea mays* ssp. *mays*) was domesticated from teosinte (*Zea mays* ssp. *parviglumis*) in the Balsas river valley of Mexico approximately 9,000 years ago. This is supported by molecular ([Matsuoka et al 2002](https://doi.org/10.1073/pnas.052125199)) and archaeological ([Piperno et al 2009](https://doi.org/10.1073/pnas.0812525106)) evidence. At a quick glance, maize and teosinte may appear to be very different. But, if you look carefully, you will find a number of similarities between the two.  

<img src="https://raw.githubusercontent.com/cjyang-work/cjyang-work.github.io/master/images/domestication1.png" width="800"/>

<em>Left</em>: maize plants, <em>middle</em>: teosinte plants, <em>right</em>: maize vs teosinte ears (not to scale).  

All of my work on maize domestication was done when I was a PhD student in John Doebley's lab at the University of Wisconsin-Madison. It was a great working with so many awesome people. Doing a PhD is never easy but it was a great experience with lots of fond memories.  

My first project was on fine-mapping a quantitative trait locus (QTL) responsible for ear internode length. This QTL was identified in [Studer and Doebley (2011)](https://doi.org/10.1534/genetics.111.126508) as *enhancer of teosinte branched1.2* (*etb1.2*) due to its interaction with the key domestication gene, *teosinte branched1* (*tb1*). The *etb1.2* QTL is located near *tb1* on chromosome 1. The work began with recombinant screening in population segregating for maize and teosinte allele at *etb1.2*, which can be laborious and you better hope your QTL of interest is in a region with high recombination rate. Fortunately, this was done before I joined John's lab and so I was able to dive right into phenotyping and fine-mapping work. Ear internode length describes how compact can the kernels be packed in a single ear, and it is measured by averaging the distance spanning 10 consecutive kernels. I recall that the measurement was not easy - most ears were not fully filled and sometimes it was impossible for find 10 consecutive kernels! We had to improvise - it was fairly easy to tell if there is only one missing kernel - but it gets harder with more missing kernels. As a good practice, we only measured the distance in the middle of the ears where the kernels are more uniform and avoided missing kernel near either ends of the measurement. It was fun to go through each ear, count and mark the kernels with a sharpie, and measured with a caliper. By relating the phenotype to the marker genotypes of the recombinant lines, we were able to narrow *etb1.2* down to a fairly small upstream region of a gene known as *ZmYAB2.1* which encodes for a YABBY transcription factor. Yay! And that was not the end, we followed up with some gene expression work and diversity sequencing to further characterize the mechanism and causative factors for the difference in ear internode length. The causative factors turned out to be rather complicated - gene expression was reduced with maize promoter over teosinte, but the presence of a large transposable element insert in the first intron appeared to abolish the gene expression completely. Wild! When comparing lines segregating for both *ZmYAB2.1* and *tb1*, we found the evidence of interaction at the level of gene expression and phenotype. We also had *in situ* hybridization results showing the spatial expression of *ZmYAB2.1* done by our amazing collaborators, Clint and Madelaine. I can't summarize this project well enough so please feel free to check out this [paper](https://doi.org/10.1534/genetics.116.194928). It is a little long but it has lots of details and work!

At about the same time, I was working with Liyan and Lexi on a project to identify additional QTL for prolificacy trait in maize and teosinte. Prolificacy is a measure of number of ears on a single lateral branch, which is one in most maize and tens in teosinte. Previous work by [Wills et al (2013)](https://doi.org/10.1371/journal.pgen.1003604) showed a large effect QTL on chromosome 1 and mapped it to a gene known as *grassy tillers1* (*gt1*) but they also found that a third of maize lines carry teosinte alleles yet exhibit non-prolific phenotype. At that point, it was thought that perhaps a different causative gene is responsible for the non-prolific phenotype in these maize lines. So, we made three F2 segregating populations between these lines (NC358, Sg18 and T8) and W22 line with teosinte introgression in *gt1* in an attempt to map the QTL for prolificacy. Surprisingly, the mapping results led us back to *gt1*. After some sequencing work, we found that these maize lines have a different haplotype in the regulatory region of *gt1* that appeared similar to teosinte allele in the original *gt1* marker. Furthermore, these maize lines turned out to be mildly prolific - they were not as prolific as teosinte but occassionally, you may find additional ears in a single lateral branch that may or may not develop to maturity. Anyway, the results were exciting and further add to the evidence of selection for multiple alleles during domestication. Please check out this [paper](https://doi.org/10.1093/jhered/esw064)!  



#### Publications
Yang CJ, Kursel LE, Studer AJ, Bartlett ME, Whipple CJ & Doebley JF (2016) A gene for genetic background in *Zea mays*: fine-mapping *enhancer of teosinte branched1.2* to a YABBY class transcription factor. Genetics 204: 1573â€“1585. [Link](https://doi.org/10.1534/genetics.116.194928)  

Yang L, Yang CJ, Cheng Q, Xue W & Doebley JF (2016) Mapping prolificacy QTL in maize and teosinte. J Hered 107: 674-678. [Link](https://doi.org/10.1093/jhered/esw064)  

Yang CJ, Samayoa LF, Bradbury PJ, Olukolu BA, Xue W, York AM, Tuholski MR, Wang W, Daskalska LL, Neumeyer MA, Sanchez-Gonzalez JJ, Romay MC, Glaubitz JC, Sun Q, Buckler ES, Holland JB & Doebley JF (2019) The genetic architecture of teosinte catalyzed and constrained maize domestication. Proc Natl Acad Sci USA 116: 5643-5652. [[Link](https://doi.org/10.1073/pnas.1820997116)]  
