---
layout: post
title: Maize Domestication
---

Maize (*Zea mays* ssp. *mays*) was domesticated from teosinte (*Zea mays* ssp. *parviglumis*) in the Balsas river valley of Mexico approximately 9,000 years ago. This is supported by molecular ([Matsuoka et al 2002](https://doi.org/10.1073/pnas.052125199)) and archaeological ([Piperno et al 2009](https://doi.org/10.1073/pnas.0812525106)) evidence. At a quick glance, maize and teosinte may appear to be very different. But, if you look carefully, you will find a number of similarities between the two.  

<img src="https://raw.githubusercontent.com/cjyang-work/cjyang-work.github.io/master/images/domestication1.png" width="800"/>
<div class="caption"><em>Left</em>: maize plants, <em>middle</em>: teosinte plants, <em>right</em>: maize vs teosinte ears (not to scale). </div>

My first project was on fine-mapping a quantitative trait locus (QTL) responsible for ear internode length. This QTL was identified in [Studer and Doebley 2011](https://doi.org/10.1534/genetics.111.126508) as *enhancer of teosinte branched1.2* (*etb1.2*) due to its interaction with the key domestication gene, *teosinte branched1* (*tb1*). The *etb1.2* QTL is located near *tb1* on chromosome 1. A recombinant inbred line carrying a teosinte introgression in *etb1.2* with maize W22 background was crossed to W22 to create an F2 population segregating for *etb1.2*. Over a thousand plants were screened for recombinations within this QTL and 45 recombinants were identified. These recombinants were further self-pollinated to create homozygous recombinant lines, which could then be evaluated in replicated field trials. This process can be laborious as you need a reasonable number of recombinants but that depends on the recombination rate around the QTL. Fortunately, this was done before I joined John's lab and so I was able to dive right into phenotyping and fine-mapping work. Ear internode length tells us how compact can the kernels be packed in a single ear, and it is measured by averaging the distance spanning 10 consecutive kernels. I recall that the measurement was not easy - most ears are not fully filled and sometimes it is impossible for find 10 consecutive kernels! We had to do what we could, and it was fairly easy to tell if there is only one missing kernel - but it gets harder with more missing kernels. As a good practice, we made sure to measure the distance in the middle of the ears where the kernels are more uniform and avoided missing kernel near either end of the measurement. It was fun to go through each ear, count and mark the kernels with a sharpie, and measured with a caliper. Anyway, if it is truly a single causative gene segregating, you would get a bimodal distribution for each parental class. In this case, it would be shorter ear internode length for maize and longer ear internode length for teosinte. By relating the marker genotypes of the recombinant lines to the phenotype, we were able to narrow *etb1.2* down to a fairly small upstream region of a gene known as *ZmYAB2.1* which encodes for a YABBY transcription factor. Yay! And that was not the end, we followed up with some gene expression work and diversity sequencing to further characterize the mechanism and causative factors for the difference in ear internode length. The causative factors turned out to be rather complicated - gene expression was reduced with maize promoter over teosinte, but the presence of a large transposable element insert in the first intron appeared to abolish the gene expression completely. Wild! When comparing lines segregating for both *ZmYAB2.1* and *tb1*, we found the evidence of interaction at the level of gene expression and phenotype. We also had *in situ* hybridization results showing the spatial expression of *ZmYAB2.1* done by our amazing collaborators, Clint and Madelaine. I don't think I can summarize this project completely here so please feel free to check out the [paper](https://doi.org/10.1534/genetics.116.194928) that came out of this project. The paper is a little long but it has lots of details and work!

#### Publications
Yang CJ, Kursel LE, Studer AJ, Bartlett ME, Whipple CJ & Doebley JF (2016) A gene for genetic background in *Zea mays*: fine-mapping *enhancer of teosinte branched1.2* to a YABBY class transcription factor. Genetics 204: 1573â€“1585. [Link](https://doi.org/10.1534/genetics.116.194928)  

Yang CJ, Samayoa LF, Bradbury PJ, Olukolu BA, Xue W, York AM, Tuholski MR, Wang W, Daskalska LL, Neumeyer MA, Sanchez-Gonzalez JJ, Romay MC, Glaubitz JC, Sun Q, Buckler ES, Holland JB & Doebley JF (2019) The genetic architecture of teosinte catalyzed and constrained maize domestication. Proc Natl Acad Sci USA 116: 5643-5652. [[Link](https://doi.org/10.1073/pnas.1820997116)]  
