---
layout: post
title: Genetic relationship lab answers
---

1.	**[Coefficient of coancestry]** Why do we see blocks of patterns in the heatmap? Is there something you can infer from the pedigree?
    + Each individual in the third generation have one full-sib.  
2.	**[Coefficient of fraternity]** Can you describe any difference between the heatmaps for coancestry versus fraternity?
    + Coefficients for coancestry are higher than fraternity, because there is not much inbreeding within the population.  
5.	**[Parentage inference]** Can you roughly count the number of recombination events in one RIL?
    + RIL01=34, RIL02=25, RIL03=34, RIL04=28, RIL05=27, RIL06=25, RIL07=28, RIL08=32, RIL09=20, RIL10=28.  
7.	**[Additive GRM]** Why are the values in the diagonals larger than the values in the off-diagonals?
    + The diagonals are the coefficients between an individual and itself so their P(IBD) are higher.  
9.	**[Dominance GRM]** Do we need D-GRM in an inbred population, and why?
    + No, because all of the off-diagonal values would be 0.  
11.	**[Compared A/D-GRM to coancestry/fraternity]** Why is there such a large discrepancy between the estimates from pedigree- vs marker-based coefficients?
    + The pedigree estimates are based on the expected probability for the given pedigree. The marker-based coefficients are closer to reality as genetic inheritance is mostly random.  
13.	**[Additive x additive GRM]** Can you explain why the AA versus A values have a quadratic/U-shaped distribution?
    + The AA coefficients are computed as the A coefficients squared, hence the quadratic pattern.  
15.	**[Splitting and combining GRM]** Why is the LOCO-GRM not perfectly correlated to the original GRM?
    + The LOCO-GRM is missing information from one chromosome.  
17.	**[Fst]** Which (sub)-population is most genetically distant from the overall population, and why?
    + Teosinte, because it has the highest Fst value of all.  
19.	**[Genome-wide Fst]** If 0.15 is out cutoff, do we have any significant marker differentiating the two (sub)-populations?
    + Yes, on chromosome 2, 7 and 9.  
21.	**[Roger’s/Nei’s distance]** Can you describe any difference or similarity between these two trees?
    + The clusters around B73 and Mo17 are quite similar between the two trees, but the overall trees are quite different.  
23.	**[PCA]** Roughly how many principal components are needed to explain 50% of the variation seen here?
    + 30.  
25.	**[PCA]** Can you describe what are each of the principal components separating?
    + PC1=Within PVP, PC2=Between PVP-CIMMYT-Teosinte-EU_Inbred, PC3=Within PVP, PC4=Between PVP-CIMMYT/Teosinte.  
27.	**[Population admixture]** Which K do you think is the right one for this dataset?
    + 2, W22 and FV2 are the only two founders.  
29.	**[GWAS]** Can you describe what is the difference in the GWAS results with and without population structure control?
    + There is a false positive on Chromosome 6.  
